### PR TITLE
fix: gcs endpoint should be independently configured with credential

### PR DESCRIPTION
**Pull request by sanadhis**
Hi, I would like to propose to decouple gcs endpoint configuration with credentials.

Use case: We're connecting to gcs via private service connect:
https://cloud.google.com/vpc/docs/private-service-connect?authuser=1&_ga=2.220919888.-993337742.1685727395#supported-apis

The access control should stay the same regardless where the bucket is accessed from.

Let me know what do you think. Appreciate any feedback!

### CODE REVIEW

**Comment by Slach** [comment]
Would you like improve your PR to add integration test for https://github.com/fsouza/fake-gcs-server?

**Review by Slach — APPROVED** [review]

**Comment by Slach** [comment]
Looks good,

did you try
```
GCS_TESTS=1 RUN_TESTS=TestIntegrationGCSWithCustomEndpoint ./test/integration/run.sh
```
??

**Comment by sanadhis** [comment]
> Looks good,
> 
> did you try
> 
> ```
> GCS_TESTS=1 RUN_TESTS=TestIntegrationGCSWithCustomEndpoint ./test/integration/run.sh
> ```
> 
> ??

Hi @Slach , thank you for some pointers.
I have resided the proposed changes and the tests. From the logs I can see that it successfully `create_remote`, `download`, `delete`, `restore`, etc. However the test didn't eventually succeeded. But I have the same result from other tests, such as `TestIntegrationS3` 

```
2024/02/19 16:10:35.760087  info docker exec clickhouse-backup clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete remote partition_backup_2361390625533319336
2024/02/19 16:10:37.018948  info 2024/02/19 15:10:35.938842  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/19 15:10:35.948011  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/19 15:10:35.948236  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:35.951193  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:35.952374  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:35.955176  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:35.971154  info SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER' logger=clickhouse
2024/02/19 15:10:35.972604  info SELECT countIf(name='type') AS is_disk_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')  logger=clickhouse
2024/02/19 15:10:35.974939  info SELECT d.path, any(d.name) AS name, any(d.type) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  INNER JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path logger=clickhouse
2024/02/19 15:10:35.983842  info done                      backup=partition_backup_2361390625533319336 duration=45ms location=remote logger=RemoveBackupRemote operation=delete
2024/02/19 15:10:35.983965  info clickhouse connection closed logger=clickhouse

2024/02/19 16:10:37.019041  info docker exec clickhouse-backup clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete local partition_backup_2361390625533319336
2024/02/19 16:10:38.339274  info 2024/02/19 15:10:37.262973  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/19 15:10:37.272927  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/19 15:10:37.272986  info SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER' logger=clickhouse
2024/02/19 15:10:37.275127  info SELECT countIf(name='type') AS is_disk_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')  logger=clickhouse
2024/02/19 15:10:37.277708  info SELECT d.path, any(d.name) AS name, any(d.type) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  INNER JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path logger=clickhouse
2024/02/19 15:10:37.281947  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:37.283787  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:37.284765  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:37.286458  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:37.296090  info done                      backup=partition_backup_2361390625533319336 duration=33ms location=local logger=RemoveBackupLocal operation=delete
2024/02/19 15:10:37.296215  info clickhouse connection closed logger=clickhouse

2024/02/19 16:10:38.351172  info testBackupSpecifiedPartitions finish
2024/02/19 16:10:38.351260  info Clean before start       
2024/02/19 16:10:38.351299  info docker exec clickhouse-backup bash -xce clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete remote TestIntegrationGCS_full_2348673835862204199
2024/02/19 16:10:38.590346  info + clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete remote TestIntegrationGCS_full_2348673835862204199
2024/02/19 15:10:38.553902  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/19 15:10:38.564154  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/19 15:10:38.564369  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:38.566992  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:38.568298  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:38.570582  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:38.581352  info clickhouse connection closed logger=clickhouse
2024/02/19 15:10:38.581393 error 'TestIntegrationGCS_full_2348673835862204199' is not found on remote storage

2024/02/19 16:10:38.590376  info docker exec clickhouse-backup bash -xce clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete local TestIntegrationGCS_full_2348673835862204199
2024/02/19 16:10:38.763957  info + clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete local TestIntegrationGCS_full_2348673835862204199
2024/02/19 15:10:38.735593  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/19 15:10:38.744409  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/19 15:10:38.744456  info SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER' logger=clickhouse
2024/02/19 15:10:38.746689  info SELECT countIf(name='type') AS is_disk_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')  logger=clickhouse
2024/02/19 15:10:38.748777  info SELECT d.path, any(d.name) AS name, any(d.type) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  INNER JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path logger=clickhouse
2024/02/19 15:10:38.751448  info clickhouse connection closed logger=clickhouse
2024/02/19 15:10:38.751483 error 'TestIntegrationGCS_full_2348673835862204199' is not found on local storage

2024/02/19 16:10:38.764012  info docker exec clickhouse-backup bash -xce clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete remote TestIntegrationGCS_increment_6730060813300271040
2024/02/19 16:10:38.955467  info + clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete remote TestIntegrationGCS_increment_6730060813300271040
2024/02/19 15:10:38.916576  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/19 15:10:38.926260  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/19 15:10:38.926452  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:38.928781  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:38.929862  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/19 15:10:38.931602  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/19 15:10:38.943932  info clickhouse connection closed logger=clickhouse
2024/02/19 15:10:38.943977 error 'TestIntegrationGCS_increment_6730060813300271040' is not found on remote storage

2024/02/19 16:10:38.955537  info docker exec clickhouse-backup bash -xce clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete local TestIntegrationGCS_increment_6730060813300271040
2024/02/19 16:10:39.145398  info + clickhouse-backup -c /etc/clickhouse-backup/config-gcs.yml delete local TestIntegrationGCS_increment_6730060813300271040
2024/02/19 15:10:39.114499  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/19 15:10:39.124084  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/19 15:10:39.124148  info SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER' logger=clickhouse
2024/02/19 15:10:39.126165  info SELECT countIf(name='type') AS is_disk_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')  logger=clickhouse
2024/02/19 15:10:39.128344  info SELECT d.path, any(d.name) AS name, any(d.type) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  INNER JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path logger=clickhouse
2024/02/19 15:10:39.130871  info clickhouse connection closed logger=clickhouse
2024/02/19 15:10:39.130903 error 'TestIntegrationGCS_increment_6730060813300271040' is not found on local storage

2024/02/19 16:10:39.145507  info docker exec clickhouse ls -1 /var/lib/clickhouse/backup/*TestIntegrationGCS*
2024/02/19 16:10:39.226175  info Drop all databases       
2024/02/19 16:10:39.235170  info docker exec minio mc ls local/clickhouse/disk_s3
2024/02/19 16:10:39.343120  info                          
2024/02/19 16:10:39.343174  info Generate test data GCS with _TestIntegrationGCS suffix
    integration_test.go:2322: 
                Error Trace:    ~/clickhouse-backup/test/integration/integration_test.go:2322
                                                        ~/clickhouse-backup/test/integration/integration_test.go:1973
                                                        ~/clickhouse-backup/test/integration/integration_test.go:1720
                Error:          Received unexpected error:
                                code: 336, message: There was an error on [127.0.0.1:9000]: Code: 336. DB::Exception: Ordinary database engine is deprecated (see also allow_deprecated_database_ordinary setting). (UNKNOWN_DATABASE_ENGINE) (version 23.8.9.54 (official build))
                Test:           TestIntegrationGCS
--- FAIL: TestIntegrationGCS (26.30s)
FAIL
FAIL    command-line-arguments  26.688s
FAIL
```

File: test/integration/config-gcs.yml [hunk 1]
```diff
@@ -3,15 +3,30 @@ general:
...
+  sync_replicated_tables: true
+  timeout: 5s
+  restart_command: "sql:SYSTEM RELOAD USERS; sql:SYSTEM RELOAD CONFIG; exec:ls -la /var/lib/clickhouse/access; sql:SYSTEM SHUTDOWN"
+  # restart_command: bash -c 'echo "FAKE RESTART"'
+  backup_mutations: true
 gcs:
   bucket: altinity-qa-test
   path: backup/{cluster}/{shard}
   object_disk_path: object_disks/{cluster}/{shard}
-  credentials_file: /etc/clickhouse-backup/credentials.json
```
Comment (Slach) [inline test/integration/config-gcs.yml]: Please, rollback changes in this file
and use different `config-gcs-facegcs.yml`

`TestIntegrationGCS` used credentials from ./test/integration/credentials.json which decrypted with https://github.com/sobolevn/git-secret

File: test/integration/config-gcs.yml [hunk 2]
```diff
@@ -3,15 +3,30 @@ general:
...
   compression_format: tar
-#  endpoint: http://gcs:8080/storage/v1/
+  endpoint: http://gcs:8080/storage/v1/
```
Comment (Slach) [inline test/integration/config-gcs.yml]: please try to use different `config-gcs-facegcs.yml`

File: test/integration/config-gcs.yml [hunk 3]
```diff
@@ -3,15 +3,30 @@ general:
...
+  skip_tables:
+    - " system.*"
+    - "INFORMATION_SCHEMA.*"
+    - "information_schema.*"
+    - "_temporary_and_external_tables.*"
   restore_schema_on_cluster: "{cluster}"
 clickhouse:
+  # wrong disk name mapping for https://github.com/Altinity/clickhouse-backup/issues/676
+  disk_mapping:
+    default-gp3: /var/lib/clickhouse
```
Comment (Slach) [inline test/integration/config-gcs.yml]: these changes have no sense

File: test/integration/run.sh [hunk 4]
```diff
@@ -17,12 +17,7 @@ fi
 export CLICKHOUSE_BACKUP_BIN="$(pwd)/clickhouse-backup/clickhouse-backup-race"
 export LOG_LEVEL=${LOG_LEVEL:-info}
 
-if [[ -f "${CUR_DIR}/credentials.json" ]]; then
```
Comment (Slach) [inline test/integration/run.sh]: Please rollback these changes

we still need to use credentials.json for test on real GCS bucket
export GSC_TESTS=${GCS_TESTS:-}
means export empty value if variable not already present

**Review by Slach — CHANGES_REQUESTED** [review]

**Comment by Slach** [comment]
> TestIntegrationGCS 
> code: 336, message: There was an error on [127.0.0.1:9000]: Code: 336. DB::Exception: Ordinary database engine is deprecated (see also allow_deprecated_database_ordinary setting). (UNKNOWN_DATABASE_ENGINE) (version 23.8.9.54 (official build))
> Test:           TestIntegrationGCS
> --- FAIL: TestIntegrationGCS (26.30s)

i don't understand why you try to fix `TestIntegrationGCS`
instead of just run
```
GCS_TESTS=1 RUN_TESTS=TestIntegrationGCSWithCustomEndpoint ./test/integration/run.sh ?
```

**Comment by Slach** [comment]
sorry I saw you have different test for GCS with custom endpoint with properly config
did you use git push --force?

**Comment by Slach** [comment]
>  code: 336, message: There was an error on [127.0.0.1:9000]: Code: 336. DB::Exception: Ordinary database engine is deprecated (see also allow_deprecated_database_ordinary setting). (UNKNOWN_DATABASE_ENGINE) (version 23.8.9.54 (official build))

it looks weird, looks like `/docker-entrypoint.d/dynamic_settings.sh` didn't run after container start

look `docker-compose_advanced.yml` for details

File: test/integration/config-gcs.yml (hunk 1 above)
Comment (sanadhis) [reply test/integration/config-gcs.yml]: oups sorry for that, I was under impression that this test is always ignored. Reverted

File: test/integration/run.sh (hunk 4 above)
Comment (sanadhis) [reply test/integration/run.sh]: Reverted

**Comment by sanadhis** [comment]
Hi @Slach thanks again for your guidance.

> sorry I saw you have different test for GCS with custom endpoint with properly config
did you use git push --force?

yes, I reverted regardless. Don't want to pollute git history.

>it looks weird, looks like /docker-entrypoint.d/dynamic_settings.sh didn't run after container start
look docker-compose_advanced.yml for details

The issue is that `dynamic_settings.sh` is missing execute permission, hence I modify the permission in my latest commit with chmod +x.

### Progress
so I made a progress on this but to be honest I don't think **fake-gcs-server** supports S3 interoperability yet. See https://github.com/fsouza/fake-gcs-server/issues/1330
The issue, that I have to make the test pass, is Clickhouse failed to connect to **fake-gcs-server**:
```
2024.02.20 10:45:31.359605 [ 91 ] {} <Error> Application: DB::Exception: Message: , bucket XXX, key disk_gcs/cluster/0/jcn/qvlyiibmadrbmeeomyonbwbykkirl, object size 4: While checking access for disk disk_gcs
```

I am not Clickhouse expert, but I am sure Clickhouse is using S3 SDK, given the `<type>s3</type>`

If I disable `QA_GCS_OVER_S3_BUCKET=XXX` in the `.env` and run `GCS_TESTS=1 RUN_TESTS=TestIntegrationGCSWithCustomEndpoint ./test/integration/run.sh`. I am advancing to this stage now:

```
...
2024/02/20 11:00:43.846247  info TestIntegrationGCSWithCustomEndpoint_increment_8299163498514678998/metadata/_issue331%2E_atomic__TestIntegrationGCSWithCustomEndpoint/_issue331%2E_atomic__TestIntegrationGCSWithCustomEndpoint.json already processed logger=resumable
2024/02/20 11:00:43.846290  info done                      backup=TestIntegrationGCSWithCustomEndpoint_increment_8299163498514678998 duration=8ms operation=upload size=227.85KiB table=_issue331._atomic__TestIntegrationGCSWithCustomEndpoint._issue331._atomic__TestIntegrationGCSWithCustomEndpoint
2024/02/20 11:00:43.857944  info done                      backup=TestIntegrationGCSWithCustomEndpoint_increment_8299163498514678998 duration=103ms operation=upload size=403.38KiB
2024/02/20 11:00:43.858131  info clickhouse connection closed logger=clickhouse

2024/02/20 12:00:44.894192  info docker exec clickhouse-backup bash -ce ls -lha /var/lib/clickhouse/backup | grep TestIntegrationGCSWithCustomEndpoint
2024/02/20 12:00:44.992739  info Delete backup            
2024/02/20 12:00:44.992796  info docker exec clickhouse-backup clickhouse-backup -c /etc/clickhouse-backup/config-gcs-custom-endpoint.yml delete local TestIntegrationGCSWithCustomEndpoint_full_8695081332883177324
2024/02/20 12:00:46.223450  info 2024/02/20 11:00:45.141930  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/20 11:00:45.152826  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/20 11:00:45.152889  info SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER' logger=clickhouse
2024/02/20 11:00:45.155353  info SELECT countIf(name='type') AS is_disk_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')  logger=clickhouse
2024/02/20 11:00:45.157869  info SELECT d.path, any(d.name) AS name, any(d.type) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  INNER JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path logger=clickhouse
2024/02/20 11:00:45.164562  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/20 11:00:45.166885  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/20 11:00:45.168029  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/20 11:00:45.169876  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/20 11:00:45.208029  info done                      backup=TestIntegrationGCSWithCustomEndpoint_full_8695081332883177324 duration=66ms location=local logger=RemoveBackupLocal operation=delete
2024/02/20 11:00:45.208195  info clickhouse connection closed logger=clickhouse

2024/02/20 12:00:46.223524  info docker exec clickhouse-backup clickhouse-backup -c /etc/clickhouse-backup/config-gcs-custom-endpoint.yml delete local TestIntegrationGCSWithCustomEndpoint_increment_8299163498514678998
2024/02/20 12:00:47.517036  info 2024/02/20 11:00:46.415699  info clickhouse connection prepared: tcp://clickhouse:9440 run ping logger=clickhouse
2024/02/20 11:00:46.425779  info clickhouse connection open: tcp://clickhouse:9440 logger=clickhouse
2024/02/20 11:00:46.425835  info SELECT value FROM `system`.`build_options` where name='VERSION_INTEGER' logger=clickhouse
2024/02/20 11:00:46.428025  info SELECT countIf(name='type') AS is_disk_type_present, countIf(name='free_space') AS is_free_space_present, countIf(name='disks') AS is_storage_policy_present FROM system.columns WHERE database='system' AND table IN ('disks','storage_policies')  logger=clickhouse
2024/02/20 11:00:46.430491  info SELECT d.path, any(d.name) AS name, any(d.type) AS type, min(d.free_space) AS free_space, groupUniqArray(s.policy_name) AS storage_policies FROM system.disks AS d  INNER JOIN (SELECT policy_name, arrayJoin(disks) AS disk FROM system.storage_policies) AS s ON s.disk = d.name GROUP BY d.path logger=clickhouse
2024/02/20 11:00:46.434657  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/20 11:00:46.436443  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/20 11:00:46.437572  info SELECT count() AS is_macros_exists FROM system.tables WHERE database='system' AND name='macros'  SETTINGS empty_result_for_aggregation_by_empty_set=0 logger=clickhouse
2024/02/20 11:00:46.439370  info SELECT macro, substitution FROM system.macros logger=clickhouse
2024/02/20 11:00:46.474979  info done                      backup=TestIntegrationGCSWithCustomEndpoint_increment_8299163498514678998 duration=60ms location=local logger=RemoveBackupLocal operation=delete
2024/02/20 11:00:46.475143  info clickhouse connection closed logger=clickhouse

2024/02/20 12:00:47.517147  info docker exec clickhouse-backup bash -ce ls -lha /var/lib/clickhouse/backup | grep TestIntegrationGCSWithCustomEndpoint
2024/02/20 12:00:47.650765  info Drop all databases       
2024/02/20 12:00:47.849617  info Download                 
2024/02/20 12:00:47.849693  info docker exec clickhouse sed -i s/<disk_gcs>/<disk_gcs_rebalanced>/g; s/<\/disk_gcs>/<\/disk_gcs_rebalanced>/g; s/<disk>disk_gcs<\/disk>/<disk>disk_gcs_rebalanced<\/disk>/g /etc/clickhouse-server/config.d/storage_configuration_gcs.xml
2024/02/20 12:00:47.933577  info sed: can't read /etc/clickhouse-server/config.d/storage_configuration_gcs.xml: No such file or directory

    integration_test.go:2156: 
                Error Trace:    ~/clickhouse-backup/test/integration/integration_test.go:2156
                                                        ~/clickhouse-backup/test/integration/integration_test.go:2017
                                                       ~/clickhouse-backup/test/integration/integration_test.go:1729
                Error:          Received unexpected error:
                                exit status 2
                Test:           TestIntegrationGCSWithCustomEndpoint
```

now I am in circle, if I enable this `QA_GCS_OVER_S3_BUCKET` Clickhouse container will fail during boot up.
Any suggestion?

**Comment by sanadhis** [comment]
The failing tests seem unrelated to my changes

File: test/integration/dynamic_settings.sh [hunk 5]
```diff
@@ -130,12 +130,14 @@ cat <<EOT > /etc/clickhouse-server/config.d/storage_configuration_gcs.xml
     <disks>
       <disk_gcs>
         <type>s3</type>
-        <endpoint>https://storage.googleapis.com/${QA_GCS_OVER_S3_BUCKET}/clickhouse_backup_disk_gcs_over_s3/${HOSTNAME}/{cluster}/{shard}/</endpoint>
-        <access_key_id>${QA_GCS_OVER_S3_ACCESS_KEY}</access_key_id>
-        <secret_access_key>${QA_GCS_OVER_S3_SECRET_KEY}</secret_access_key>
+        <endpoint>http://gcs:8080/${QA_GCS_OVER_S3_BUCKET}/disk_gcs/{cluster}/{shard}/</endpoint>
```
Comment (Slach) [inline test/integration/dynamic_settings.sh]: i think this is unnecessary, please rollback

File: test/integration/integration_test.go [hunk 6]
```diff
@@ -1720,6 +1720,15 @@ func TestIntegrationGCS(t *testing.T) {
 	runMainIntegrationScenario(t, "GCS", "config-gcs.yml")
 }
 
+func TestIntegrationGCSWithCustomEndpoint(t *testing.T) {
+	if isTestShouldSkip("GCS_TESTS") {
+		t.Skip("Skipping GCS integration tests...")
+		return
+	}
+	//t.Parallel()
+	runMainIntegrationScenario(t, "GCS", "config-gcs-custom-endpoint.yml")
```
Comment (Slach) [inline test/integration/integration_test.go]: let's change `GCS` to something like `GCS_EMULATOR`
to avoid disk rebalancing feature

File: test/integration/integration_test.go [hunk 7]
```diff
@@ -2119,9 +2127,8 @@ func checkObjectStorageIsEmpty(t *testing.T, r *require.Assertions, remoteStorag
 			checkRemoteDir("total 0", "ftp", "bash", "-c", "ls -lh /home/vsftpd/test_backup/backup/")
 		}
 	}
-	//todo check gcs backup is empty
 	if remoteStorageType == "GCS" {
```
Comment (Slach) [inline test/integration/integration_test.go]: let's change to `== "GCS_EMULATOR"` here

**Review by Slach — CHANGES_REQUESTED** [review]

**Comment by Slach** [comment]
`<disk_gcs_rebalanced>`  used for `GCS` for some corner cases, lets use `GCS_EMULATOR` as workaround

File: test/integration/dynamic_settings.sh (hunk 5 above)
Comment (sanadhis) [reply test/integration/dynamic_settings.sh]: reverted, the only change in this file is it's now executable.

**Comment by sanadhis** [comment]
> `<disk_gcs_rebalanced>` used for `GCS` for some corner cases, lets use `GCS_EMULATOR` as workaround

Thanks @Slach , the integration test has passed now 😄 
```
2024/02/20 14:26:38.413323  info docker exec clickhouse ls -1 /var/lib/clickhouse/backup/*TestIntegrationGCSWithCustomEndpoint*
2024/02/20 14:26:38.543924  info Drop all databases       
2024/02/20 14:26:38.698641  info docker exec gcs sh -c ls -lh /data/altinity-qa-test/
--- PASS: TestIntegrationGCSWithCustomEndpoint (84.78s)
PASS
ok      command-line-arguments  85.163s

```

**Review by Slach — APPROVED** [review]
will try to fix failed tests later

**Comment by sanadhis** [comment]
@Slach is there planned release soon?

**Comment by Slach** [comment]
check https://github.com/Altinity/clickhouse-backup/releases/tag/v2.4.32

**Comment by sanadhis** [comment]
> check https://github.com/Altinity/clickhouse-backup/releases/tag/v2.4.32

Superb, thanks a lot 👍
